### ISSUE TITLE
Flat hierarchy resolution

### ISSUE THREAD

**Issue by daniel-brosche** [jacebrowning/gitman#128]
i'm thinking about to implement a flat hierarchy resolution strategy like bower. I think it would be good extension to gitman as long it is configurable in the update or install method. What do you think?

**Comment by jacebrowning** [comment]
`gitman list` shows the full tree after installation. Do you have any thoughts on how this might be stored for later reinstall?

**Comment by daniel-brosche** [comment]
I'm not sure if understand the question. Basically my idea is in addition to the recursive resolution stretegies where each repository stores it dependencies in a subdirectory similar to submodules is to realize a flat resolution strategy where all dependencies are resolved in one folder. Recursive resolution of all dependencies but the storage in one folder, that's the idea. I have scenarios where this resolution strategy is quite benefitial. Of course resolution conflicts needs to be handled too and will happen more often instead of the recursive storage strategy. Does this answer your question?

**Comment by jacebrowning** [comment]
I think I understand. After running `gitman install`, instead of getting something like

```
.gitman.yml
vendor/
  dep_a/
    .gitman.yml
    vendor/
      dep_c
  dep_b/
    .gitman.yml
    vendor/
      dep_c
```

you would get something like:

```
.gitman.yml
vendor/
  dep_a/
  dep_b/
  dep_c/
```

where somehow you resolve `dep_c` to a single version. Does that sound right?

**Comment by daniel-brosche** [comment]
Correct, that's the idea... This resolution strategy is similar to the dependency management of bower or the Perforce import Streams functionality. I have worked in large projects with dozens of dependencies which are resolved in this way. 
My idea is to switch the resolution strategy by a --flat argument which can be passed to the `gitman update` resp. `gitman install `operation. 
Something like `gitman install --flat`.

**Comment by jacebrowning** [comment]
This feature sounds useful. It might even be worth having a `flat: <bool>` setting in the configuration file because you would want dependencies to be checked out the same on every machine.

**Comment by daniel-brosche** [comment]
In my experience this resolution strategy is much more helpful in large projects with many dependencies and a more complex dependency graph. There are also situations where it is quite harmful to import distinct dependency multiple times (dep_c v1.0 and dep_c v2.0). I would start to implement this feature soon because I need this feature for a complex embeeded system project.

**Comment by daniel-brosche** [comment]
After some prototyping, I think the best way to implement this feature is not introduce a seperate command option `--flat` and instead extend the configuration file to specify the flat dependency resolution.
Therefore I think, we could introduce a  newstring field `resolve` which is by default `recursive` and can be specified to `flat`. In the future we could add further strategies.
